File: Striver A2Z/Step 6 - Learn LinkedList [Single LL, Double LL, Medium, Hard Problems] /Lec 5 - Hard Problems of LL/reverse-nodes-in-k-group.cpp

// https://leetcode.com/problems/reverse-nodes-in-k-group/description/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    int lengthLL(ListNode*head){
        int len = 0;
        if(!head) return len;
        while(++len && head) head = head->next;
        return (len-1);
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        
        if(!head) return head;

        int currLen = lengthLL(head);
        if(currLen < k) return head;
        int count = k;
        ListNode*prev = NULL, *curr = head, *fwd = NULL;
        while(curr && count--){
            fwd = curr->next;
            curr->next = prev;
            prev = curr;
            curr = fwd;
        }

        if(fwd) head->next = reverseKGroup(fwd, k);
        return prev;

    }
};

File: Neetcode-150-js/Linked List/reverse-nodes-in-k-group.js

// https://leetcode.com/problems/reverse-nodes-in-k-group/description/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */

var lengthLL = function(head){
    let length = 0;
    if(!head) return length;
    while(head){
        head = head.next;
        length++;
    }
    return length;
}

var reverseKGroup = function(head, k) {
    if(!head || k === 0) return head;
    let length = lengthLL(head);
    if(length < k) return head;

    let curr = head, fwd = null, prev = null;
    let count = 0;
    while(curr && count < k){
        fwd = curr.next;
        curr.next = prev;
        prev = curr;
        curr = fwd;
        count++;
    }

    if(fwd) head.next = reverseKGroup(curr, k);
    return prev;
};